### DISCUSSION TITLE
How to dynamically generate UUID in curl command (like macros) before sending http request?

### DISCUSSION THREAD

**Discussion by morozover** [curl/curl#13768] [Q&A] [answered]
I want to send POST http request via curl in Windows command line app.
There is `x-client-request-id` header in request
I want to create curl command that dynamically generate UUID for this header

Something like that:
```
curl -X POST -H "x-client-request-id: {{UUID}}" -d '{"a": 1}' https:/example.com
```

This doesn't work:
```
${uuid}
{{$uuid}}
```

**Comment by dfandrich** [comment]
I'm not that familiar with the Windows shell, but where is the value for {{UUID}} supposed to be coming from?

**Comment by jay** [comment, marked as answer]
> curl -X POST -H "x-client-request-id: {{UUID}}" -d '{"a": 1}' https:/example.com

[Unnecessary use of curl -X](https://daniel.haxx.se/blog/2015/09/11/unnecessary-use-of-curl-x/)

[-d, --data](https://curl.se/docs/manpage.html#-d) sets the request method to POST

Your quoting is wrong for the Windows command shell. I'm not sure if it would work in powershell. Windows command shell uses double quotes to quote arguments so you'd have to do it like `-d "{\"a\": 1}"`. Also as a timesaver curl has a [--json](https://curl.se/docs/manpage.html#--json) option to set the headers to content json, so you may be able to do `--json "{\"a\": 1}"`.

For `{{UUID}}` again it depends on the shell. In the Windows command shell you probably mean `%UUID%` like
~~~
C:\>set UUID=test

C:\>echo %UUID%
test
~~~

Familiarize yourself with the shell you are using and how to quote arguments and expand variables. Some of the examples in the curl manual use unix shell quoting and expansion which is different from Windows.

edit: I missed the question in your title. You can use `uuidgen` in Windows or `uuidgen -r` in Linux. I think you will need Visual Studio for uuidgen in Windows.
~~~
C:\> uuidgen > uuid

C:\> set /p UUID=<uuid

C:\> echo %UUID%
5eb4123f-0038-4517-961c-8aab422a4f21
~~~

**Comment by morozover** [comment]
@jay 
I see that there are third party utilities in command shell/Power shell for generating UUID

But could you please clarify, has curl **built-in** macros for generating UUID in request itself?

As in Postman:
![image](https://github.com/curl/curl/assets/17902666/cd0b51c1-fe04-4cac-8927-62046d5171f5)

**Reply by bagder** [reply]
> has curl built-in macros for generating UUID in request itself?

No it does not. It seems a little random to me to offer that.

**Reply by bagder** [reply]
Apparently postman has quite a few [variables](https://learning.postman.com/docs/tests-and-scripts/write-scripts/variables-list/). Maybe we should consider providing a few of them?

**Reply by morozover** [reply]
I would be great

**Reply by bagder** [reply]
I decided to test out this as an idea and see what people think:  https://github.com/curl/curl/pull/13813
